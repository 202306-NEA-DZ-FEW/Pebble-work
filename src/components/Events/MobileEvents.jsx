import React, { useState, useEffect } from "react";
import Calendar from "@/components/Filter/Calendar";
import EventCardLeft from "@/components/Events/EventCardLeft";
import styles from "@/styles/Events.module.css";
import {
    collection,
    getDocs,
    onSnapshot,
    query,
    where,
} from "firebase/firestore";

import FilterByType from "@/components/Filter/FilterByType";
import FirestoreLocation from "@/components/Filter/FirestoreLocation";

import { db } from "@/util/firebase";

const EventsPage = (user) => {
    // State variables

    const [selectedTypes, setSelectedTypes] = useState([]);
    const [inputValue1, setInputValue1] = useState("");
    const [filteredTypes, setFilteredTypes] = useState([]);
    const [events, setEvents] = useState([]);
    const [filteredEvents, setFilteredEvents] = useState([]);
    const [CalendarEvents, setCalendarEvents] = useState([]);
    const [selectedDate, setSelectedDate] = useState(null);
    const [resetLocation, setResetLocation] = useState(false);
    const [resetDays, setResetDays] = useState(false);

    const handleLocationInputChange = (value) => {
        setInputValue1(value);
    };

    const filterEventsByLocation = (location) => {
        if (!location) {
            setFilteredEvents([]);
            return () => {};
        }

        const eventsCollectionRef = collection(db, "events");
        const q = query(eventsCollectionRef, where("location", "==", location));

        const unsubscribe = onSnapshot(q, (querySnapshot) => {
            const matchingEvents = querySnapshot.docs.map((doc) => {
                return { id: doc.id, ...doc.data() };
            });
            setFilteredEvents(matchingEvents);
        });

        return unsubscribe;
    };
    useEffect(() => {
        const unsubscribe = filterEventsByLocation(inputValue1);

        return () => {
            unsubscribe();
        };
    }, [inputValue1]);

    const checkEvents = async (selectedDate) => {
        const q = query(
            collection(db, "events"),
            where("date", "==", selectedDate)
        );

        try {
            const querySnapshot = await getDocs(q);
            const filteredEvents = querySnapshot.docs.map((doc) => doc.data());
            setCalendarEvents(filteredEvents);
            console.log(CalendarEvents);
            setSelectedDate(selectedDate);
        } catch (error) {
            console.error("Error getting filtered events: ", error);
        }
    };

    // Fetch events from Firebase
    useEffect(() => {
        const fetchEvents = async () => {
            const eventsCollectionRef = collection(db, "events");
            const eventsSnapshot = await getDocs(eventsCollectionRef);
            const eventsData = eventsSnapshot.docs.map((doc) => {
                return { id: doc.id, ...doc.data() };
            });

            setEvents(eventsData);
        };

        fetchEvents();
    }, []);

    useEffect(() => {
        const applyFilters = () => {
            let filteredEvents = events;

            // Apply type filter
            if (filteredTypes.length > 0) {
                filteredEvents = filteredEvents.filter((event) =>
                    filteredTypes.includes(event.type)
                );
            }

            // Apply location filter
            if (inputValue1) {
                filteredEvents = filteredEvents.filter(
                    (event) => event.location === inputValue1
                );
            }

            // Apply date filter
            if (selectedDate) {
                filteredEvents = filteredEvents.filter(
                    (event) => event.date === selectedDate
                );
            }

            setFilteredEvents(filteredEvents);
        };

        applyFilters();
    }, [events, selectedDate, inputValue1, filteredTypes]);

    const resetEvents = () => {
        setSelectedTypes([]);
        setInputValue1("");
        setSelectedDate(null);
        setFilteredTypes([]);
        setFilteredEvents([]);
        setCalendarEvents([]);
        setResetLocation(true);
        setResetDays([]);
    };
    const handleInputDelete = () => {
        setInputValue1("");
    };
    return (
        <>
            <main
                className={` flex flex-col justify-center items-center pb-12`}
            >
                <div>
                    <h1>Welcome, {user.name}!</h1>
                    <p>This is the events page</p>
                </div>
                <button
                    onClick={resetEvents}
                    className={` w-[52px] bg-blue-400 text-white text-[10px] hover:bg-blue-500 md:text-[12px] rounded-[4px] h-[16px] sm:w-[72.23px] sm:h-[25.5px] ml-auto  mr-2`}
                >
                    All events
                </button>
                <div
                    className={`flex flex-col-reverse sm:flex sm:flex-row-reverse sm:items-center sm:justify-evenly sm:gap-8 sm:h-full sm:w-full`}
                >
                    <div
                        className={`h-[400px] pb-[140px] md:pb-[140px] md:w-[480px] sm:h-[1000px] ${styles.information}`}
                    >
                        <ul className={` flex flex-col items gap-2 `}>
                            {filteredEvents.map((event, index) => {
                                return (
                                    <EventCardLeft
                                        eventId={event.id}
                                        key={event.id}
                                        title={event.title}
                                        type={event.type}
                                        image={event.image}
                                        location={event.location}
                                        description={event.description}
                                        organizer={event.organizer}
                                        time={event.time}
                                        date={event.date}
                                    />
                                );
                            })}
                            {(inputValue1 || filteredTypes.length > 0) &&
                                filteredEvents.length === 0 && (
                                    <p className='text-red-500 text-center'>
                                        No events found for this date and
                                        location
                                    </p>
                                )}

                            {selectedDate && CalendarEvents.length === 0 && (
                                <p className='text-red-500 text-center'>
                                    No events found for this date
                                </p>
                            )}
                        </ul>
                    </div>
                    <div className='flex bg-white z-10 flex-row items-center justify-between sm:flex sm:flex-col sm:items-center text-black sm:gap-7'>
                        <div className='sm:flex s:flex-col sm:items-center sm:justify-center'>
                            <button className='sm:hidden'>Dates</button>

                            <div
                                className={`border border-black rounded-[8px] z-10 bg-white sm:bg-transparent`}
                            >
                                <Calendar
                                    resetDays={resetDays}
                                    checkEvents={checkEvents}
                                />
                            </div>
                        </div>
                        <div className='h-66'>
                            <FirestoreLocation
                                onInputChange={handleLocationInputChange}
                                resetLocation={resetLocation}
                                setResetLocation={setResetLocation}
                                onInputDelete={handleInputDelete}
                            />
                        </div>
                        <FilterByType
                            setFilteredTypes={setFilteredTypes}
                            selectedTypes={selectedTypes}
                            setSelectedTypes={setSelectedTypes}
                        />
                    </div>
                </div>
            </main>
        </>
    );
};

export default EventsPage;
