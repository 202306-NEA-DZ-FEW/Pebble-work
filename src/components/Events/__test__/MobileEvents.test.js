import React from "react";

import MobileEvents from "../MobileEvents";
import { render, cleanup } from "@testing-library/react";
afterEach(cleanup);
describe("MobileEvents", () => {
    it("renders correctly", () => {
        const user = { name: "Test User" };

        const { asFragment } = render(<MobileEvents user={user} />);

        expect(asFragment()).toMatchSnapshot();
    });
});
