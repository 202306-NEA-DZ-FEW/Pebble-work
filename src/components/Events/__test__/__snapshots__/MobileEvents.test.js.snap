// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsPage component renders correctly 1`] = `
<main
  className="flex flex-col justify-center items-center pb-12"
  style={
    {
      "color": "#749D60",
      "textShadow": "0px 1px 2px rgba(10, 10, 10, 0.8)",
    }
  }
>
  <div>
    <h1>
      Welcome, 
      !
    </h1>
    <p>
      This is the events page
    </p>
  </div>
  <div
    className="flex flex-col-reverse sm:flex sm:flex-row-reverse sm:items-center sm:justify-evenly sm:gap-8 sm:h-full sm:w-full"
  >
    <div
      className="md:w-[480px] sm:h-[800px]"
    >
      <ul
        className=" flex flex-col items gap-2 "
      >
        <div
          className="flex justify-center gap-8 text-black"
        />
      </ul>
    </div>
    <div
      className="flex z-10 flex-row items-center justify-between sm:flex sm:flex-col sm:items-center text-black sm:gap-7"
    >
      <div
        className="sm:flex s:flex-col sm:items-center sm:justify-center"
      >
        <div
          className="shadow-md rounded-[8px] z-10 bg-[#B4CD93]"
        >
          <div
            className="flex flex-col items-center"
          >
            <div
              className="flex flex-col-reverse items-center justify-center xl:w-[333px] xl:h-[243px] sm:w-[222px] sm:h-[180px]"
            >
              <div
                className="flex justify-center items-center gap-2"
              >
                <button
                  className="flex items-center"
                  onClick={[Function]}
                >
                  <img
                    alt="Previous"
                    className="xl:w-[40px] rotate-180 xl:h-[40px] w-[25px] md:h-[25px]"
                    data-nimg="1"
                    decoding="async"
                    height={50}
                    loading="lazy"
                    onError={[Function]}
                    onLoad={[Function]}
                    src="/icons/next.svg"
                    style={
                      {
                        "color": "transparent",
                      }
                    }
                    width={50}
                  />
                </button>
                <div
                  className="flex flex-col items-center gap-3"
                >
                  <div
                    className="calendar sm:h-[140px] xl:h-[200px] grid grid-cols-7 gap-1"
                  >
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      S
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      M
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      T
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      W
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      T
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      F
                    </div>
                    <div
                      className="sm:text-[12px] font-[500] w-[25px] md:ml-1 sm:ml-2 xl:text-[17px]"
                    >
                      S
                    </div>
                    <div
                      className="emptyDay"
                    />
                    <div
                      className="emptyDay"
                    />
                    <div
                      className="emptyDay"
                    />
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      1
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      2
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      3
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      4
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      5
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      6
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      7
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      8
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      9
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      10
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      11
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      12
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      13
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      14
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      15
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      16
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      17
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      18
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      19
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      20
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      21
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      22
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      23
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      24
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      25
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      26
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      27
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      28
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      29
                    </button>
                    <button
                      className="calendarDay  xl:text-[17px] md:w-[10px] xl:w-[18px] text-center sm:text-[11px] rounded"
                      onClick={[Function]}
                    >
                      30
                    </button>
                  </div>
                  <h2
                    className="text-xl xl:w-[200px] sm:w-[165px] flex justify-center gap-4 font-bold sm:text-[14px] xl:text-[18px]"
                  >
                    <div
                      className="text-xl bg-transparent font-bold sm:text-[14px] xl:text-[18px] dropdown"
                    >
                      <button
                        onClick={[Function]}
                      >
                        ^
                        November
                      </button>
                    </div>
                    <div
                      className="text-xl bg-transparent font-bold sm:text-[14px] xl:text-[18px] dropdown"
                    >
                      <button
                        onClick={[Function]}
                      >
                        ^
                        2023
                      </button>
                    </div>
                  </h2>
                </div>
                <button
                  className="flex items-center"
                  onClick={[Function]}
                >
                  <img
                    alt="Next"
                    className="xl:w-[40px] xl:h-[40px] w-[25px] md:h-[25px]"
                    data-nimg="1"
                    decoding="async"
                    height={50}
                    loading="lazy"
                    onError={[Function]}
                    onLoad={[Function]}
                    src="/icons/next.svg"
                    style={
                      {
                        "color": "transparent",
                      }
                    }
                    width={50}
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="h-66 shadow-md"
      >
        <button
          className="btn bg-[#B4CD93] hover:bg-gray-400 sm:text-center flex lg:items-center sm:gap-0 gap-2 sm:justify-center lg:w-[281px] lg:h-[52px] sm:w-[149px] sm:h-[30px] text-white font-[500] lg:text-[18px] lg:tracking-[0.10px] sm:text-[12px] sm:px-4 sm:py-2 rounded-lg"
          onClick={[Function]}
        >
          Location
        </button>
        <dialog
          className="modal rounded-tr rounded-br"
          id="my_modal_1"
        >
          <div
            className="modal-box"
          >
            <div
              className="flex justify-center"
            >
              <input
                className="mt-2 p-2 border rounded"
                onChange={[Function]}
                placeholder="Search city..."
                type="text"
                value=""
              />
            </div>
            <h3
              className="font-bold text-lg mb-0"
            >
              cities 
            </h3>
            <div
              className=" mt-2  "
            >
              <ul
                className="list-none information h-[200px]"
              />
              <div
                className="modal-action absolute top-7 right-3 mt-0"
              >
                <form
                  method="dialog"
                >
                  <button
                    className=""
                    onClick={[Function]}
                  >
                    <svg
                      fill="currentColor"
                      fillRule="evenodd"
                      height={20}
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 1024 1024"
                      width={20}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
                      />
                    </svg>
                  </button>
                </form>
              </div>
            </div>
          </div>
        </dialog>
      </div>
      <div
        className=" shadow-md"
      >
        <div>
          <button
            className="sm:hidden"
            onClick={[Function]}
          >
            Interests
          </button>
        </div>
      </div>
    </div>
  </div>
</main>
`;
